src/Servers/Kestrel/Transport.Quic/src/Internal/QuicStreamContext.cs

```diff
@@ -566,15 +566,17 @@ public override async ValueTask DisposeAsync()
                 && !_serverAborted
                 && _shutdownReadReason == null
                 && _shutdownWriteReason == null;
+        }
```

Comment:
Null out `_stream` inside the lock, I assume `DisposeCore` is safe outside of the lock.

Reply:
I moved it because `DisposeAsync` can't happen inside a lock.

The stream shouldn't be used anymore because `_processingTask` is awaited. The one place it could be used, and `_stream` is accessed in a lock, is `Abort`.

I've changed `Abort` so that it takes a local copy of `_stream` and then checks it for null.

Reply:
So basically, we need to be able to call `_stream.Abort()` while `_stream.DisposeAsync()` is running? And anything else would be invalid since middleware has already completed.

The only problem is that Abort() could in theory still be in the middle of executing while we return the context to the pool because it doesn't do any synchronization. This could be bad if Abort() does something like setting _serverAborted after the QuicStreamContext has already been pushed and poped from the pool and "reset". I don't think Abort() can be thread safe without acquiring the _shutdownLock.

The other thing to make sure of is that we null `_stream` out before it enters the pool. DisposeAsync() is always awaited by Http3Stream before it calls Dispose() which may pool this QuicStreamContext, so we're good. Maybe putting it at the start of `Dispose()` before the `if (!_connection.TryReturnStream(this))` would make this more obvious.

Reply:
I've changed:
- `DisposeAsync` to call `stream.DisposeAsync` before the lock, and moved setting `_stream` to null inside the lock again. I think I like it more now, but I don't think this makes a huge difference as draining will already have happened by other code awaiting `WaitForWritesClosedAsync`. Because draining has already happened, `stream.DisposeAsync` should complete immediately.
- `Abort` now checks for null `_stream` and uses lock when determining whether it should run.